curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,418 @@
+        notFull_.wait(guard);
+    }
+    compactingInodes_.push_back(inodeKey);
+    auto task = std::bind(&S3CompactWorkQueueImpl::CompactChunks, this,
+                          inodeStorage, inodeKey, pinfo, copysetNode);
+    queue_.push_back(std::move(task));
+    notEmpty_.notify_one();
+}
+
+std::function<void()> S3CompactWorkQueueImpl::Dequeue() {
+    std::unique_lock<std::mutex> guard(mutex_);
+    while (queue_.empty() && running_.load(std::memory_order_acquire)) {
+        notEmpty_.wait(guard);
+    }
+    std::function<void()> task;
+    if (!queue_.empty()) {
+        task = std::move(queue_.front());
+        queue_.pop_front();
+        compactingInodes_.pop_front();
+        notFull_.notify_one();
+    }
+
+    return task;
+}
+
+void S3CompactWorkQueueImpl::ThreadFunc() {
+    while (running_.load(std::memory_order_acquire)) {
+        Task task(Dequeue());
+        if (task) {
+            task();
+        }
+    }
+}
+
+std::vector<uint64_t> S3CompactWorkQueueImpl::getNeedCompact(
+    const ::google::protobuf::Map<uint64_t, S3ChunkInfoList>& s3chunkinfoMap) {
+    std::vector<uint64_t> needCompact;
+    int maxChunkPerIndex;
+    int fragmentThreshold;
+    LOG_IF(FATAL, !conf_->GetIntValue("s3compactwq.max_chunks_per_index",
+                                      &maxChunkPerIndex));
+    LOG_IF(FATAL, !conf_->GetIntValue("s3compactwq.fragment_threshold",
+                                      &fragmentThreshold));
+    for (const auto& item : s3chunkinfoMap) {
+        if (needCompact.size() >= maxChunkPerIndex) {
+            break;
+        }
+        if (item.second.s3chunks_size() > fragmentThreshold) {
+            needCompact.push_back(item.first);
+        }
+    }
+    return needCompact;
+}
+
+void S3CompactWorkQueueImpl::deleteObjs(const std::vector<std::string>& objs) {
+    for (const auto& obj : objs) {
+        const Aws::String aws_key(obj.c_str(), obj.size());
+        s3Adapter_->DeleteObject(aws_key);  // don't care success or not
+    }
+}
+
+std::list<struct S3CompactWorkQueueImpl::Node>
+S3CompactWorkQueueImpl::buildValidList(const S3ChunkInfoList& s3chunkinfolist,
+                                       uint64_t inodeLen) {
+    std::list<struct S3CompactWorkQueueImpl::Node> validList;
+    std::function<void(uint64_t, uint64_t, uint64_t, uint64_t, uint64_t,
+                       uint64_t)>
+        addToValidList;
+    addToValidList = [&](uint64_t begin, uint64_t end, uint64_t chunkid,
+                         uint64_t compaction, uint64_t chunkoff,
+                         uint64_t zero) {
+        if (end > inodeLen) end = inodeLen;
+        for (auto it = validList.begin(); it != validList.end();) {
+            uint64_t nodeBegin = it->begin;
+            uint64_t nodeEnd = it->end;
+            uint64_t nodeChunkid = it->chunkid;
+            uint64_t nodeCompaction = it->compaction;
+            uint64_t nodeChunkoff = it->chunkoff;
+            bool nodeZero = it->zero;
```

Comment:
what's the difference between perameter `zero` and `nodeZero`, line 135 emplace takes `zero` but line 157 takes `nodeZero`, and meanwhile Node's constructor takes bool value.

Reply:
"zero" is a attr of "message S3ChunkInfo", will be added by huyao's PR, indicating the data is all zero or not. when truncated (expand), it will appear.
"node" prefix means attr from current node in validlist, i will add some comments to help people understand.